### PR TITLE
[Feat][metadata] RocksDB MetadataStore Support ReadOnly

### PR DESCRIPTION
**Pull request by liangyuanpeng**
### Motivation

1. RocksDB MetadataStore Support ReadOnly  

Using RocksDB meatadata outside of pulsar for  analyze.

2. Fix NPE on open RocksDb MetadataStore failed.

3. MetadataStore support config when running with standalone.

The doc say rocksdb's configuration file path `metadataStoreConfigPath` is configurable in the `standalone.conf` and in fact, it's not working. This PR fixed it.

### Modifications

Add System properties of `pulsar.metadata.rocksdb.readonly` for enable readonly.

Let Standalone use `config.getMetadataStoreConfigPath()`  for metadataStore configuration.

They are small enough so i commit it together.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

*(example:)*
  -  Running rocksdb MetadataStore with ReadOnly and put data, Got the error 
 - Running rocksdb MetadataStore with ReadOnly and delete data,Got the error 

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-metadata/src/test/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStoreTest.java [hunk 1]
```diff
@@ -35,11 +35,64 @@
...
+import static org.testng.Assert.assertEquals;
 
 @Slf4j
 public class RocksdbMetadataStoreTest {
 
+    MetadataStore store;
+
+    Path readonlyDataDir;
+    @BeforeClass
+    public void setup() throws Exception {
```
Comment (Jason918) [inline pulsar-metadata/src/test/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStoreTest.java:49]: This method does not apply to other old test method. Please use a new test class or call it manually in the new tests.

File: pulsar-metadata/src/test/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStoreTest.java [hunk 2]
```diff
@@ -35,11 +35,64 @@
...
+        readonlyDataDir = Files.createTempDirectory("RocksDBMetadataReadOnlyTest");
+
+         store = MetadataStoreFactory.create("rocksdb:"+readonlyDataDir, MetadataStoreConfig.builder().build());
+         // put init data to rocksdb
+        store.put("/test","hello".getBytes(), Optional.of(-1L)).get();
+        store.close();
```
Comment (Jason918) [inline pulsar-metadata/src/test/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStoreTest.java:55]: Can we use readonly mode when normal store is not closed?

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java [hunk 3]
```diff
@@ -233,12 +240,15 @@ private RocksdbMetadataStore(String metadataURL, MetadataStoreConfig metadataSto
...
 
+        readonly = "true".equalsIgnoreCase(
+                System.getProperty(ROCKSDB_READONLY, "false"));
         db = openDB(dataPath.toString(), metadataStoreConfig.getConfigFilePath());
 
         this.optionSync = new WriteOptions().setSync(true);
         this.optionDontSync = new WriteOptions().setSync(false);
         this.optionCache = new ReadOptions().setFillCache(true);
         this.optionDontCache = new ReadOptions().setFillCache(false);
+        dbStateLock = new ReentrantReadWriteLock();
```
Comment (Jason918) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java:251]: Any reason to move this here?

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java [hunk 4]
```diff
@@ -233,12 +240,15 @@ private RocksdbMetadataStore(String metadataURL, MetadataStoreConfig metadataSto
             throw new MetadataStoreException("Fail to create RocksDB file directory", e);
         }
 
+        readonly = "true".equalsIgnoreCase(
```
Comment (Jason918) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java:243]: I am not sure if it's a good idea to use system property to set up this config. It's not easy for user to know this config.

**Review by Jason918 — CHANGES_REQUESTED** [review]

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java (hunk 3 above)
Comment (liangyuanpeng) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java:251]: You would got the NPE when init rocksdb failed,

https://github.com/apache/pulsar/pull/17078/files#diff-d8d2e43e5722098642f3e3b650804e38c12c15cb9f5b92ed2f5a612618498d11R259

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java (hunk 4 above)
Comment (liangyuanpeng) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java:243]: My first impl version is using rocksdb config file (metadata config file) and the `rocksdb config file` is more used to work for rocksdb core, so i change it to system property for this config.

Do you have any suggestion?

File: pulsar-metadata/src/test/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStoreTest.java (hunk 1 above)
Comment (liangyuanpeng) [reply pulsar-metadata/src/test/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStoreTest.java:49]: Sure, would be change to just call it for readonly mode.

**Comment by tisonkun** [comment]
Closed as no consensus.

**Pull request closed without merging**